File: viewer-ui/src/api/catalog.ts
import axios from 'axios';

const api = axios.create({
  baseURL: '',
  headers: {
    'Content-Type': 'application/json',
  },
});

export interface CatalogueArtworkMap {
  poster?: string;
  banner?: string;
  thumbnail?: string;
}

export interface CatalogueEpisodeVariant {
  episode_id: string;
  language: string;
  title: string;
  duration_sec: number;
  synopsis?: string;
}

export interface CatalogueEpisode {
  content_group: string;
  episode_number: number;
  title: string;
  duration_sec: number;
  synopsis?: string;
  languages: string[];
  artworks: CatalogueArtworkMap;
  variants: CatalogueEpisodeVariant[];
}

export interface CatalogueTrailer {
  episode_id: string;
  title: string;
  language: string;
  duration_sec: number;
  artworks: CatalogueArtworkMap;
}

export interface CatalogueSeason {
  season_number: number;
  episodes: CatalogueEpisode[];
}

export interface CatalogueShow {
  id: string;
  title: string;
  slug: string;
  section: string;
  categories: string[];
  synopsis?: string;
  artworks: CatalogueArtworkMap;
  trailers: CatalogueTrailer[];
  seasons: CatalogueSeason[];
  total_episodes: number;
}

export interface PublishedCatalogue {
  version: number;
  generated_at: string;
  generated_by: string;
  total_shows: number;
  total_episodes: number;
  sections: Record<string, CatalogueShow[]>;
}

export interface SearchResponse {
  query: string;
  matched_shows_count: number;
  matched_episodes_count: number;
  results: CatalogueShow[];
}

export const fetchCatalogue = async (): Promise<PublishedCatalogue> => {
  const { data } = await api.get('/catalog');
  return data;
};

export const searchCatalogue = async (params: {
  q?: string;
  category?: string;
  language?: string;
  section?: string;
}): Promise<SearchResponse> => {
  const { data } = await api.get('/catalog/search', { params });
  return data;
};

export const fetchShowDetail = async (idOrSlug: string): Promise<CatalogueShow> => {
  const { data } = await api.get(`/catalog/shows/${idOrSlug}`);
  return data;
};

File: backend/app/schemas/catalog.py
from datetime import datetime
from typing import List, Dict, Optional, Any
from pydantic import BaseModel, ConfigDict

class CatalogueArtworkMap(BaseModel):
    poster: Optional[str] = None
    banner: Optional[str] = None
    thumbnail: Optional[str] = None

class CatalogueEpisodeVariant(BaseModel):
    episode_id: str
    language: str
    title: str
    duration_sec: int
    synopsis: Optional[str] = None

class CatalogueEpisodeItem(BaseModel):
    content_group: str
    episode_number: int
    title: str
    duration_sec: int
    synopsis: Optional[str] = None
    languages: List[str]
    artworks: CatalogueArtworkMap = CatalogueArtworkMap()
    variants: List[CatalogueEpisodeVariant] = []

class CatalogueTrailerItem(BaseModel):
    episode_id: str
    title: str
    language: str
    duration_sec: int
    artworks: CatalogueArtworkMap = CatalogueArtworkMap()

class CatalogueSeasonItem(BaseModel):
    season_number: int
    episodes: List[CatalogueEpisodeItem] = []

class CatalogueShowItem(BaseModel):
    id: str
    title: str
    slug: str
    section: str
    categories: List[str]
    synopsis: Optional[str] = None
    artworks: CatalogueArtworkMap = CatalogueArtworkMap()
    trailers: List[CatalogueTrailerItem] = []
    seasons: List[CatalogueSeasonItem] = []
    total_episodes: int = 0

class PublishedCatalogue(BaseModel):
    version: int
    generated_at: str
    generated_by: str
    total_shows: int
    total_episodes: int
    sections: Dict[str, List[CatalogueShowItem]]

    model_config = ConfigDict(from_attributes=True)

class CatalogSearchResponse(BaseModel):
    query: str
    matched_shows_count: int
    matched_episodes_count: int
    results: List[CatalogueShowItem]
